plugins/languages/treesitter/treesitter.nix

```diff
@@ -6,210 +6,303 @@
+          This package is required to include grammars to install.
+        '';
+      }
+      {
+        assertion = ((cfg.grammarPackages != [ ]) || cfg.grammarPackages == "all") -> cfg.package != null;
+        message = ''
+          Nixvim (plugins.treesitter): You have enabled `grammarPackages`, but `plugins.treesitter.package` is `null`.
+          This package is required to choose which grammars to install.
+        '';
+      }
+      {
+        assertion = !cfg.nixGrammars -> cfg.gccPackage != null;
+        message = ''
+          Nixvim (plugins.treesitter): You have disabled `nixGrammars`, but `plugins.treesitter.gccPackage` is `null`.
+          This package is required to build the grammars at runtime.
+        '';
+      }
+      {
+        assertion = !cfg.nixGrammars -> cfg.nodejsPackage != null;
+        message = ''
+          Nixvim (plugins.treesitter): You have disabled `nixGrammars`, but `plugins.treesitter.nodejsPackage` is `null`.
+          This package is required to build the grammars at runtime.
+        '';
+      }
+    ];
 
-        incremental_selection =
-          if cfg.incrementalSelection.enable then
-            {
-              enable = true;
-              keymaps = {
-                init_selection = cfg.incrementalSelection.keymaps.initSelection;
-                node_incremental = cfg.incrementalSelection.keymaps.nodeIncremental;
-                scope_incremental = cfg.incrementalSelection.keymaps.scopeIncremental;
-                node_decremental = cfg.incrementalSelection.keymaps.nodeDecremental;
-              };
-            }
-          else
-            null;
-
-        indent = if cfg.indent then { enable = true; } else null;
-
-        ensure_installed = if cfg.nixGrammars then [ ] else cfg.ensureInstalled;
-        ignore_install = cfg.ignoreInstall;
-        parser_install_dir = cfg.parserInstallDir;
-      } // cfg.moduleConfig;
-    in
-    mkIf cfg.enable {
-      extraConfigLua =
-        (optionalString (cfg.parserInstallDir != null) ''
-          vim.opt.runtimepath:append("${cfg.parserInstallDir}")
-        '')
-        + ''
-          require('nvim-treesitter.configs').setup(${helpers.toLuaObject tsOptions})
-        ''
-        + (optionalString (cfg.languageRegister != { }) ''
-          __parserFiletypeMappings = ${helpers.toLuaObject cfg.languageRegister}
+    extraConfigLua =
+      # NOTE: It MUST be at the beginning of runtimepath. Otherwise the parsers from Neovim itself
+      # is loaded that may not be compatible with the queries from the 'nvim-treesitter' plugin.
+      (optionalString (cfg.settings.parser_install_dir != null) ''
+        do
+          vim.opt.runtimepath:prepend("${cfg.settings.parser_install_dir}")
+        end
+      '')
+      + ''
+        do
+          require('nvim-treesitter.configs').setup(${helpers.toLuaObject cfg.settings})
+        end
```

Comment:
I think it's fine if the `do .. end` block encloses the _entire_ `extraConfigLua` definition.

i.e. one `do .. end` block, not multiple

Reply:
Alright, just saw another example in repo do it this way and thought it was preferred. Can change it back.

Reply:
I haven't looked too closely at other examples, so IDK if other maintainers have a preference.

To me, it makes sense to just have one scope unless there's a specific reason not to.

Reply:
I don't think we really need a `do .. end` block here as there is no issue with local variable leaking.
As we don't use that for most plugins, I would just not have it.
(Have I missed something) ?

Reply:
We have the local `__parserFiletypeMappings`. Not a big deal as it's double underscored, but I think having a `do .. end` is better